### ISSUE TITLE
lint: gcassert doesn't seem to work (at least in some cases)

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#121312]
I applied this diff on master:
```
diff --git a/pkg/sql/distsql_physical_planner.go b/pkg/sql/distsql_physical_planner.go
index 0ba6c9cb680..17ed5379f58 100644
--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -382,6 +382,7 @@ func checkExprForDistSQL(expr tree.Expr) error {
        if expr == nil {
                return nil
        }
+       //gcassert:noescape
        v := distSQLExprCheckVisitor{}
        tree.WalkExprConst(&v, expr)
        return v.err
diff --git a/pkg/testutils/lint/gcassert_paths.txt b/pkg/testutils/lint/gcassert_paths.txt
index 2bdfac932b5..06463998d9a 100644
--- a/pkg/testutils/lint/gcassert_paths.txt
+++ b/pkg/testutils/lint/gcassert_paths.txt
@@ -7,6 +7,7 @@ kv/kvpb
 kv/kvserver/intentresolver
 kv/kvserver/rangefeed
 roachpb
+sql
 sql/catalog/descs
 sql/colcontainer
 sql/colconv
```
and expected `TestGCAssert` to fail, but it didn't. I confirmed by looking at output of `go build` that `v` escapes to heap:
```
pkg/sql/distsql_physical_planner.go:386:2: v escapes to heap:
```
I then confirmed than running `gcassert ./pkg/sql` does produce the expected output:
```
pkg/sql/distsql_physical_planner.go:386:	v := distSQLExprCheckVisitor{}: v escapes to heap:
```
It must be that using GCAssert via the linter is broken.

I debugged this a bit, and probably https://github.com/jordanlewis/gcassert/commit/55a956e1897b7ac5aa247760f955ff208b8946f5 was insufficient for our needs. I saw this as `gcassert.go:227: absPath`
![Screenshot 2024-03-28 at 3 50 51 PM](https://github.com/cockroachdb/cockroach/assets/5135385/3e908f60-79f8-49e8-9cc1-0d675d0533bb)
whereas `directiveMap` had this
![Screenshot 2024-03-28 at 3 51 16 PM](https://github.com/cockroachdb/cockroach/assets/5135385/5aa0e1e9-5e33-4940-a8e1-3c2d9afa2e04)
In other words, looks like the path we use to invoke `gcassert` contains `testutils` when performing compilation but not when parsing gcassert directives from the source code.

I'm wondering whether GCAssert linter currently doesn't work at all.

cc @rickystewart 

Jira issue: CRDB-37180

**Comment by rickystewart** [comment]
There are small problems on both the `gcassert` side and the `lint` side. Doesn't seem like any legitimate failures have squeaked through though (luckily). I'll get started with the `gcassert` patch.
